### PR TITLE
feat/ Improve gateway nonce architecture

### PR DESCRIPTION
**Pull request by CrimsonJacket**
**Before submitting this PR, please make sure**:

- [x] Your code builds clean without any errors or warnings
- [x] You are using approved title ("feat/", "fix/", "docs/", "refactor/")

**A description of the changes proposed in the pull request**:
[sc-24553] - Nonce Architecture Improvement

**Tests performed by the developer**:
Remove nonce management mechanism in `GatewayEVMAMM`.
Refactor how nonce is being managed by `EVMNonceManager`. 
`/nonce` endpoint and `EVMNonceManager.getNonce()` now returns the nonce of the last successful transaction.
Introduced `EVMNonceManager.getNextNonce()` to determine the next sequential nonce to use if the `nonce` parameter is not provided
Introduced `EVMNonceManager.isValidNonce()` to determine validity of the `nonce` parameter if provided in a request

**Tips for QA testing**:
Ensure that `amm-arb` strategy works as intended.
Using `manual-tests` issue multiple concurrent `/amm/trade` requests. All requests should be successful.
A possible

### CODE REVIEW

File: gateway/src/services/error-handler.ts [hunk 1]
```diff
@@ -86,6 +86,7 @@ export const SWAP_PRICE_EXCEEDS_LIMIT_PRICE_ERROR_CODE = 1008;
 export const SWAP_PRICE_LOWER_THAN_LIMIT_PRICE_ERROR_CODE = 1009;
 export const SERVICE_UNITIALIZED_ERROR_CODE = 1010;
 export const UNKNOWN_CHAIN_ERROR_CODE = 1011;
+export const INVALID_NONCE_ERROR_CODE = 1012;
```
Comment (vic-en) [inline gateway/src/services/error-handler.ts:89]: We should update documentation for this.

**Comment by PtrckM** [comment]
- getting an error when executing a arbitrage opportunity

image
- binance -> pangolin


```
2022-04-13 05:00:34,526 - 21095 - GatewayEVMAMM - ERROR - Error submitting BUY swap order to pangolin on avalanche for 0.200000000000000 WAVAX-USDC.e 77.470819700450000.
Traceback (most recent call last):
  File "/home/ubuntu/improvenonce/hummingbot/connector/gateway_EVM_AMM.py", line 512, in _create_order
    order_result: Dict[str, Any] = await GatewayHttpClient.get_instance().amm_trade(
TypeError: amm_trade() missing 1 required positional argument: 'nonce'
2022-04-13 05:00:34,529 - 21095 - hummingbot.core.event.event_reporter - EVENT_LOG - {"timestamp": 1649826034.0, "order_id": "buy-WAVAX-USDC.e-1649826034525993", "order_type": "OrderType.LIMIT", "event_name": "MarketOrderFailureEvent", "event_source": "pangolin_avalanche_avalanche"}
```

image
- binance -> uniswap 

```
2022-04-13 05:06:50,002 - 21957 - GatewayEVMAMM - ERROR - Error submitting BUY swap order to uniswap on kovan for 0.008000000000000 WETH-USDC 10.331978318750000.
Traceback (most recent call last):
  File "/home/ubuntu/improvenonce/hummingbot/connector/gateway_EVM_AMM.py", line 512, in _create_order
    order_result: Dict[str, Any] = await GatewayHttpClient.get_instance().amm_trade(
TypeError: amm_trade() missing 1 required positional argument: 'nonce'
2022-04-13 05:06:50,005 - 21957 - hummingbot.core.event.event_reporter - EVENT_LOG - {"timestamp": 1649826410.0, "order_id": "buy-WETH-USDC-1649826410001997", "order_type": "OrderType.LIMIT", "event_name": "MarketOrderFailureEvent", "event_source": "uniswap_ethereum_kovan"}
```

- logs, configs, csv and sqlite:
[amm.zip](https://github.com/CoinAlpha/hummingbot/files/8478349/amm.zip)

File: gateway/src/services/error-handler.ts (hunk 1 above)
Comment (CrimsonJacket) [reply gateway/src/services/error-handler.ts:89]: Which documentation are you talking about? If you're talking about Swagger, I dont see any error definitions anywhere.

Comment (vic-en) [reply gateway/src/services/error-handler.ts:89]: https://www.notion.so/hummingbot/API-Specs-ee319e13355a4906af6ab30bfe1496ea

File: hummingbot/core/gateway/gateway_http_client.py [hunk 2]
```diff
@@ -398,8 +399,9 @@ async def amm_trade(
             "side": side.name,
             "amount": f"{amount:.18f}",
             "limitPrice": str(price),
-            "nonce": nonce
         }
+        if nonce is not None:
+            request_payload["nonce"] = str(nonce)
```
Comment (vic-en) [inline hummingbot/core/gateway/gateway_http_client.py:404]: This should remain an int. Otherwise, it'll fail validation as defined above.

**Review by vic-en — CHANGES_REQUESTED** [review]

**Comment by PtrckM** [comment]
Issue reported above are now fixed. ✅ 

> * getting an error when executing a arbitrage opportunity
>

**Review by PtrckM — APPROVED** [review]
LGTM! 
Test summary:

Install feature branch without any issues
Create an amm_arbitrage strategy using uniswap and binance without any issue
Start the strategy and wait for the preliminary check to finish
Closely monitor during trade filled event on both side confirm no issue
Enable concurrent order and execute trade without any issue
Try couple of different pairs and no issue
Repeat the step but using pangolin connector and confirm it is behaving as expected
Run manual-test /amm/trade endpoint and confirmed working
Run the trade endpoint multiple times to issue concurrent orders and confirmed working as well.

File: gateway/src/services/error-handler.ts (hunk 1 above)
Comment (CrimsonJacket) [reply gateway/src/services/error-handler.ts:89]: Included the new Error Codes in the Notion doc

**Review by vic-en — APPROVED** [review]
All changes LGTM

**Pull request merged**